### PR TITLE
GMM compatibility and usability fixes

### LINKED ISSUE #321: Mask convergence warnings in AutoGMM
Currently, if individual runs of GMM in sklearn did not converge, a warning is thrown. This can happen many times over multiple inits. I think we should suppress this specific warning, since if it is thrown we already increase the regularization anyway. Could have an option to throw a warning that says "increasing regularization" if necessary.

### LINKED ISSUE #324: Depreciation warning in AutoGMM
## Actual Behavior
Importing AutoGMM throws the following warning

## Full Traceback
```
FutureWarning: The sklearn.mixture.gaussian_mixture module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.mixture. Anything that cannot be imported from sklearn.mixture is now part of the private API.
 warnings.warn(message, FutureWarning)

```

## Your Environment
* Sklearn version 0.22

### LINKED ISSUE #331: Make cosine similarity handling in AutoGMM less fragile
Currently, errors are thrown when trying to use cosine similarity on a dataset that contains at least one 0-vector. 

I think we should have AGMM skip cosine similarity in this case, potentially throwing a warning.

### PR DESCRIPTION
**Pull request by caseypw**
Closes #321
Closes #324 
Closes #331

#### What does this implement/fix? Explain your changes.
- Fixes cosine similarity handling when a row is 0, in this case, cosine similarity is skipped and a warning is thrown
- Fixes deprecation warnings for Gaussian mixture by doing proper import 
- Masks convergence warnings internally in AutoGMM since regularization is increased when first attempts don't converge

### CODE REVIEW

File: graspy/cluster/autogmm.py [hunk 1]
```diff
@@ -377,86 +373,6 @@ def __init__(
         self.verbose = verbose
         self.selection_criteria = selection_criteria
         self.max_agglom_size = max_agglom_size
-        self.n_jobs = n_jobs
-
-    def _fit_cluster(self, X, y, params):
```
Comment (bdpedigo) [inline graspy/cluster/autogmm.py:382]: @caseypw can you explain? It looks like perhaps you were not on an up-to-date version of master before making all of your changes, can you revise to only reflect the changes you want to make?

Comment (caseypw) [reply graspy/cluster/autogmm.py:382]: My bad, I'm not sure where these changes came from. Will get proper version of this file.

**Comment by caseypw** [comment]
Not really sure why it is failing netlify checks @bdpedigo @j1c 

Here is the error output:

9:18:13 PM: /opt/build/repo/graspy/cluster/gclust.py:docstring of graspy.cluster.GaussianCluster:38: WARNING: Bullet list ends without a blank line; unexpected unindent.
9:18:13 PM: /opt/build/repo/graspy/cluster/gclust.py:docstring of graspy.cluster.GaussianCluster:64: WARNING: Unexpected indentation.
9:18:13 PM: WARNING: error while formatting arguments for graspy.cluster.AutoGMMCluster: 'function' object has no attribute '__mro__'
9:18:13 PM: WARNING: [numpydoc] While processing docstring for 'graspy.cluster.AutoGMMCluster'
9:18:13 PM: Exception occurred:
9:18:13 PM:   File "/opt/buildhome/python3.6/lib/python3.6/site-packages/numpydoc/docscrape.py", line 625, in __init__
9:18:13 PM:     raise ValueError("Expected a class or None, but got %r" % cls)
9:18:13 PM: ValueError: Expected a class or None, but got <function AutoGMMCluster at 0x1501273abd90>

**Comment by bdpedigo** [comment]
> Not really sure why it is failing netlify checks @bdpedigo @j1c
> 
> Here is the error output:
> 
> 9:18:13 PM: /opt/build/repo/graspy/cluster/gclust.py:docstring of graspy.cluster.GaussianCluster:38: WARNING: Bullet list ends without a blank line; unexpected unindent.
> 9:18:13 PM: /opt/build/repo/graspy/cluster/gclust.py:docstring of graspy.cluster.GaussianCluster:64: WARNING: Unexpected indentation.
> 9:18:13 PM: WARNING: error while formatting arguments for graspy.cluster.AutoGMMCluster: 'function' object has no attribute '**mro**'
> 9:18:13 PM: WARNING: [numpydoc] While processing docstring for 'graspy.cluster.AutoGMMCluster'
> 9:18:13 PM: Exception occurred:
> 9:18:13 PM: File "/opt/buildhome/python3.6/lib/python3.6/site-packages/numpydoc/docscrape.py", line 625, in **init**
> 9:18:13 PM: raise ValueError("Expected a class or None, but got %r" % cls)
> 9:18:13 PM: ValueError: Expected a class or None, but got <function AutoGMMCluster at 0x1501273abd90>

I think by using a decorator you changed the definition of AutoGMMCluster to a function, is my hunch. Try removing to see if that fixes. May not be able to use the decorator in this way, im not sure.

**Comment by bdpedigo** [comment]
also will need a test for the cosine similarity thing

**Comment by bdpedigo** [comment]
@caseypw really need to get on this if you want this merged before end of class (in addition to all the other stuff you said you were going to do). Generally don't review PRs until tests/builds are all passing. Also see that I requested at least a test for the cosine similarity changes.

**Comment by caseypw** [comment]
Checks pass @bdpedigo @j1c

**Comment by caseypw** [comment]
> also will need a test for the cosine similarity thing

I added a test for the warning raised in response to cosine similarity, is that what you needed here @bdpedigo ?

File: graspy/cluster/autogmm.py [hunk 2]
```diff
@@ -17,12 +17,15 @@
...
-from sklearn.mixture.gaussian_mixture import (
+from sklearn.mixture._gaussian_mixture import (
     _compute_precision_cholesky,
     _estimate_gaussian_parameters,
 )
 from sklearn.model_selection import ParameterGrid
+from sklearn.utils._testing import ignore_warnings
+from sklearn.exceptions import ConvergenceWarning
 from joblib import Parallel, delayed
+import warnings
```
Comment (bdpedigo) [inline graspy/cluster/autogmm.py:28]: @caseypw  is this import necessary

**Comment by bdpedigo** [comment]
@caseypw edge case to think about, how does this behave when there is one "affinity" and it's cosine, but the matrix has a 0 vector?

**Comment by bdpedigo** [comment]
@caseypw  can you add a note to the docs about the cosine similarity thing, under the options for affinity? Maybe just something along the lines of "Note that cosine similarity can only work when all of the rows are not the zero vector, if the input matrix has a zero row, cosine similarity will be skipped and a warning will be thrown". And then make sure it renders/looks okay in the netlify build

**Comment by bdpedigo** [comment]
@caseypw do you need the ignore_warnings decorator in all 3 places? I'd try to get by with only decorating one of those functions if possible

File: graspy/cluster/autogmm.py (hunk 2 above)
Comment (caseypw) [reply graspy/cluster/autogmm.py:28]: Yes because in line 518 I throw a warning for the cosine similarity case. Did we still want a warning for that?

Comment (bdpedigo) [reply graspy/cluster/autogmm.py:28]: ah ok that's fine

**Comment by bdpedigo** [comment]
@caseypw the other two comments above were not addressed yet, correct? just tag me for review again when they are

**Comment by caseypw** [comment]
> @caseypw edge case to think about, how does this behave when there is one "affinity" and it's cosine, but the matrix has a 0 vector?

I had AutoGMM raise a ValueError and end the execution to address this

**Comment by caseypw** [comment]
> @caseypw can you add a note to the docs about the cosine similarity thing, under the options for affinity? Maybe just something along the lines of "Note that cosine similarity can only work when all of the rows are not the zero vector, if the input matrix has a zero row, cosine similarity will be skipped and a warning will be thrown". And then make sure it renders/looks okay in the netlify build

@bdpedigo Should I address this in the other PR I am going to make that has the tutorial notebooks? To address #333

**Comment by caseypw** [comment]
All checks pass @bdpedigo

**Comment by bdpedigo** [comment]
> > @caseypw can you add a note to the docs about the cosine similarity thing, under the options for affinity? Maybe just something along the lines of "Note that cosine similarity can only work when all of the rows are not the zero vector, if the input matrix has a zero row, cosine similarity will be skipped and a warning will be thrown". And then make sure it renders/looks okay in the netlify build
> 
> @bdpedigo Should I address this in the other PR I am going to make that has the tutorial notebooks? To address #333

No, don't worry about it for the tutorial notebooks

**Comment by caseypw** [comment]
All checks pass @bdpedigo @j1c 
I added a line under affinity, below the lines "If a list/array ..."

Is it ready to be merged?

**Pull request merged**